### PR TITLE
Keras Bigdl backend example

### LINKED ISSUE #2180: Enrich optimize method parameters for keras bigdl backend
take a look at method `to_bigdl_optim_method`

### PR DESCRIPTION
**Pull request by hkvision**
## What changes were proposed in this pull request?

Map keras losses and add ut on python side.
Map keras optimization methods with parameters.
Add corresponding imdb example.

## How was this patch tested?

Jenkins

## Related links or issues (optional)

Fix #2180

### CODE REVIEW

File: pyspark/bigdl/examples/keras/imdb_bigdl_backend.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+X_train, y_train, X_test, y_test = load_imdb()
+model = build_keras_model()
+
+model.compile(loss='binary_crossentropy',
+              optimizer='adam',
+              metrics=['accuracy'])
+
+# add this line of code to use BigDL backend alternatively
+model = with_bigdl_backend(model)
```
Comment (jason-dai) [inline pyspark/bigdl/examples/keras/imdb_bigdl_backend.py]: Shall we do below instead? @zhichao-li 
```
model = with_bigdl_backend(model)
model.compile()
model.fit()
`

File: pyspark/bigdl/examples/keras/imdb_bigdl_backend.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+# Alternative of the example imdb_cnn_lstm.py with BigDL backend
+# You can essentially employ the same code written in Keras 1.2.2 and add the following line
+# after building the model and before training the model.
+#     model = with_bigdl_backend(model)
+
+from bigdl.keras.backend import *
+from bigdl.examples.keras.imdb_cnn_lstm import *
+
+X_train, y_train, X_test, y_test = load_imdb()
+model = build_keras_model()
```
Comment (zhichao-li) [inline pyspark/bigdl/examples/keras/imdb_bigdl_backend.py]: rename to `kmodel`?

Comment (hkvision) [reply pyspark/bigdl/examples/keras/imdb_bigdl_backend.py]: I think maybe `model` is better here? As for illustration purpose, we can use the same code in keras and add a line to run in BigDL. Thus, from the perspective of users, they don't need to distinguish kmodel or bmodel?

**Comment by hkvision** [comment]
TODO: Add documentation for bigdl backend
TODO: Add tests for loss

**Comment by hkvision** [comment]
jenkins passed http://172.168.2.101:8080/view/PR-V-P/job/BigDL-PR-V/774/

**Comment by hkvision** [comment]
@zhichao-li have a review?

File: pyspark/bigdl/keras/backend.py [hunk 3]
```diff
@@ -20,18 +20,21 @@
...
+        redire_spark_logs()
+        show_bigdl_info_logs()
         self.bmodel = DefinitionLoader.from_kmodel(kmodel)
         WeightLoader.load_weights_from_kmodel(self.bmodel, kmodel)  # share the same weight.
         self.criterion = OptimConverter.to_bigdl_criterion(kmodel.loss)
         self.optim_method = OptimConverter.to_bigdl_optim_method(kmodel.optimizer)
         self.metrics = OptimConverter.to_bigdl_metrics(kmodel.metrics) if kmodel.metrics else None
 
-    def evaluate(self, x, y, batch_size=32, sample_weight=None, is_distributed=False):
+    def evaluate(self, x, y, batch_size=32, sample_weight=None, is_distributed=True):
```
Comment (zhichao-li) [inline pyspark/bigdl/keras/backend.py]: any reason for  this?

File: pyspark/bigdl/keras/optimization.py [hunk 4]
```diff
@@ -38,26 +40,80 @@ def to_bigdl_metrics(metrics):
...
+            return bcriterion.CategoricalCrossEntropy()
+        elif kloss == "mse" or kloss == "mean_squared_error" or kloss == mse:
             return bcriterion.MSECriterion()
-        elif kloss == "binary_crossentropy":
+        elif kloss == "binary_crossentropy" or kloss == binary_crossentropy:
             return bcriterion.BCECriterion()
-        elif kloss == "mae" or kloss == "mean_absolute_error":
+        elif kloss == "mae" or kloss == "mean_absolute_error" or kloss == mae:
             return bcriterion.AbsCriterion()
+        elif kloss == "hinge" or kloss == hinge:
```
Comment (zhichao-li) [inline pyspark/bigdl/keras/optimization.py]: any difference?

Comment (hkvision) [reply pyspark/bigdl/keras/optimization.py]: Yes. The former is a string and the latter is a function.

File: pyspark/bigdl/keras/backend.py (hunk 3 above)
Comment (hkvision) [reply pyspark/bigdl/keras/backend.py]: I think default using distributed fashion is better?
1. For BigDL, our feature towards Keras is that we can do training in a distributed way.
2. Seems that distributed is better supported. For example, evaluation/validation doesn't support local mode. If user use the default false, he/she will get an error.

Comment (zhichao-li) [reply pyspark/bigdl/keras/backend.py]: Ideally, we should be able to run keras code without changes. so let's turn the default value to be false to meet that requirement.  We can fix the not supported error in a separated PR.

File: pyspark/bigdl/keras/optimization.py (hunk 4 above)
Comment (zhichao-li) [reply pyspark/bigdl/keras/optimization.py]: ok

File: pyspark/bigdl/keras/backend.py [hunk 5]
```diff
@@ -109,15 +112,15 @@ def fit(self, x, y=None, batch_size=32, nb_epoch=10, verbose=1, callbacks=None,
         if validation_split != 0.:
             unsupport_exp("validation_split")
         bopt = self.__create_optimizer(x=x,
-                                y=y,
-                                batch_size=batch_size,
-                                nb_epoch=nb_epoch,
-                                validation_data=validation_data,
-                                is_distributed=is_distributed)
+                                       y=y,
```
Comment (zhichao-li) [inline pyspark/bigdl/keras/backend.py]: not sure what's the difference here? the code on server using "space" but you are using "tab" ?

Comment (hkvision) [reply pyspark/bigdl/keras/backend.py]: Emmm... just keep the parameters aligned... for the sake of better visualization... doesn't change the code anyway.

File: pyspark/bigdl/keras/backend.py (hunk 3 above)
Comment (hkvision) [reply pyspark/bigdl/keras/backend.py]: OK. I will fix it.

File: pyspark/bigdl/examples/keras/imdb_bigdl_backend.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+# Define a model in Keras 1.2.2 and compile
+keras_model = build_keras_model()
+keras_model.compile(loss='binary_crossentropy',
+                    optimizer='adam',
+                    metrics=['accuracy'])
+
+# Add this line of code to use BigDL backend alternatively
+model = with_bigdl_backend(keras_model)
+
+model.fit(X_train, y_train, batch_size=32, nb_epoch=2,
```
Comment (zhichao-li) [inline pyspark/bigdl/examples/keras/imdb_bigdl_backend.py]: add `evaluate` and `predict` here as well?

Comment (hkvision) [reply pyspark/bigdl/examples/keras/imdb_bigdl_backend.py]: Added.

**Comment by zhichao-li** [comment]
LGTM

**Comment by hkvision** [comment]
jenkins passed http://172.168.2.101:8080/view/PR-V/job/BigDL-PR-V/861/
